### PR TITLE
edit links [stub- do not merge]

### PR DESCRIPTION
**Pull request by olivierdalang**
Regarding issue #11 , I started too see if it's easy to implement. For now it's a very quick hack, do not expect much from it.

There are now three widget classes :
AddEditWidgetWrapper
AddAnotherWidgetWrapper
EditSelectedWidgetWrapper

The thing is that for the Edit link, we need to pass the ID. That's relatively easy, since the ID is the value of the <select> element. So one needs to pass the url with '**id**' as a placeholder for the ID. It is then replaced by the current value of the <select> using inline javascript. If '**id**' isn't in the url, the id gets appended to the end.

This works pretty well, as the popup mechanics works also for the change form.

For now, the edit link is visible even if there's no selection, because I didn't find how to get javascript's events for when the select changes (I'm using django-autocomplete-light, so it's not a regular select). And for now, the javascript is inline...

What do you think ?

Bests,

Olivier

### CODE REVIEW

File: django_addanother/widgets.py [hunk 1]
```diff
@@ -69,5 +77,17 @@ def render(self, name, value, *args, **kwargs):
...
             'add_icon': self.add_icon,
+            'edit_related_url': self.edit_related_url,
+            'edit_icon': self.edit_icon,
         }
         return mark_safe(render_to_string(self.template, context))
+
+
+class AddAnotherWidgetWrapper(AddEditWidgetWrapper):
+    def __init__(self, widget, add_related_url, add_icon=None):
+        print('hehe ;) im here')
```
Comment (jpic) [inline django_addanother/widgets.py]: Hehe, finding your way ? ;)

**Comment by jpic** [comment]
I'd like to propose renaming AddEditAnotherWidgetWrapper to RelatedWidgetWrapper or something, do we really need both AddAnotherWidgetWrapper and EditAnotherWidgetWrapper ?

**Comment by olivierdalang** [comment]
Hehe forgot that print statement. 

Ok for the renaming and no we don't really need the two classes (one can just provide no url if he don't want the add/edit button).

But with those classes it's backward compatible and slightly less verbose. 

I don't really mind on this so I'll do as you like. 

Besides I got a strange issue with the popups and grappelli : everything works fine except that the pop-up aren't closed. Once the form is submitted the pop-up displays a blank page and it's title is "Popup closing..."
Any idea why it's not closing? 
Could it be a jQuery conflict? (I don't understand how one is supposed to deal with jQuery in django/grappelli)

File: django_addanother/templates/django_addanother/related_widget_wrapper.html [hunk 2]
```diff
@@ -2,9 +2,22 @@
 <div class="related-widget-wrapper">
     {{ widget }}
 
+    {% if edit_related_url %}
+    &nbsp;
```
Comment (jonashaag) [inline django_addanother/templates/django_addanother/related_widget_wrapper.html]: Why is this necessary?

**Comment by jonashaag** [comment]
Cool! Suggestions:
- Maybe we could replace the `__id__` hackery with simply annotating each `<option>` with its edit URL.
- Don't hide the edit button if no valid selection is made. I'd say disable it instead. This has two advantages: 1) It's less distracting. 2) It makes it clear that the feature exists but it's disabled.
- Maybe even show the edit button even if no edit URL is given, for the reasons stated above?

Otherwise, looks good to me, and I'm in favour of merging this. (I'll have to do a more thorough code review before that though)

**Comment by olivierdalang** [comment]
About annotating each `<option>`, I'm not really sure on how it would work. Here I use django-autocomplete-light, so I'm not sure it's even possible ?

Ok for not hiding when there's no valid selection, but I think it's better not to display it when there's no URL given. Here I have an use case where I want the user to be able to easily edit a relation, but not to add one. I guess the opposite could exist as well.

**Comment by jpic** [comment]
The annotation idea is clean, however, shouldn't we try to stick to the way django admin does it for better compatibility with the ecosystem ?

**Comment by jonashaag** [comment]
Absolutely. I wasn't aware of the fact that this is the Django Way (TM). Maybe add a comment about that fact then.

**Comment by olivierdalang** [comment]
@jpic @jonashaag 
I'm not aware of the django way of doing it, is it (by coincidence!) that `__id__` hack, or are you talking about something else ?

I made a new commit, now the "edit" stays, but disabled, when no object is selected.

As I suggested, there is no icon at all if no url is given, as in this case, it means no matter what the user does, there won't be a way to button becomes editable.

What do you think ?

File: django_addanother/static/django_addanother/img/icon-changelink-disabled.svg [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<svg width="13" height="13" viewBox="0 0 1792 1792" xmlns="http://www.w3.org/2000/svg">
+  <path fill="#b8b8b8" d="M491 1536l91-91-235-235-91 91v107h128v128h107zm523-928q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zm-54-192l416 416-832 832h-416v-416zm683 96q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"/>
+</svg>
```
Comment (jpic) [inline django_addanother/static/django_addanother/img/icon-changelink-disabled.svg]: Why don't we use django admin's here ?

File: django_addanother/static/django_addanother/edit_related.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+        }
+
+    });
+
+
+    $('.related-widget-wrapper select').change();
+
+});
+
```
Comment (jpic) [inline django_addanother/static/django_addanother/edit_related.js]: Can't we reuse django.contrib.admin's JS here like for add-another ?

**Comment by olivierdalang** [comment]
The first commit removes custom js and uses django admin's javascripts.

It almost works, but I get `Uncaught ReferenceError: interpolate is not defined` when trying to use `dismissChangeRelatedObjectPopup` defined in `/static/admin/js/admin/RelatedObjectLookups.js`

I'm quite lost when it comes to javascript dependencies in django. I think it has to do with `jsi18n` not being loaded, but I'm not sure how to add this to the widget's medias. Could someone help ?

Also, I changed the test project in the second commit (it has custom models that will allow a more adapted test project), don't know if you care to merge this too or not ?

**Comment by jpic** [comment]
> Also, I changed the test project in the second commit (it has custom models that will allow a more adapted test project), don't know if you care to merge this too or not ?

I wanted to try your branch but that apparently broke tox: 

```
ERROR: InvocationError: '/home/jpic/env/src/django-addanother/.tox/py27-django19/bin/py.test -vv --cov django_addanother --create-db --strict -r fEsxXw test_project/tests.py'
```

**Comment by olivierdalang** [comment]
I deleted the test.py file by mistake, just restored it... but tox still fails here. I don't understand what's happening (never used tox before).

If you prefer I can also revert the whole test case change (it's just that I needed some regular foreign keys to test).

**Comment by jpic** [comment]
Just check the commands travis runs to find out how tox is used here. Correct me if I'm wrong @jonashaag but we'll need another test to be added in this file to prevent regressions on this new feature you're contributing :)

To fix the interpolate error here, we just need to add `/admin/jsi18n/` to the widget's Media. Is there any way to reverse this url perhaps ?

Keep up the great work :guitar:

**Comment by olivierdalang** [comment]
I don't get what the error is with tox... I googled a bit, but didn't find anything useful. Any hint would be very welcome, I'm new to tox, and I probably miss something obvious about what's causing the error.

The build seems to start correctly (there are some flake8 warnings), but stops with `ERROR: InvocationError: '/home/travis/build/jonashaag/django-addanother/.tox/checkqa/bin/flake8 --show-source --exclude tests --max-complexity=5 --ignore=D203,E501,I101 django_addanother'` which doesn't give much information...

I tried locally (installed tox using `pip install tox`, then simply running `tox`) but I get also an `InvocationError`, this time with a log. But the log is also cryptic, and the error seems related to local problems : 
`py34-django18-0.log`

```
actionid: py34-django18
msg: getenv
cmdargs: ['/home/olivier/virtualenvs/toxtests/bin/python3', '-m', 'virtualenv', '--python', '/home/olivier/virtualenvs/toxtests/bin/python3.4', 'py34-django18']
env: {'SHELL': '/bin/bash', 'SSH_TTY': '/dev/pts/0', '_': '/home/olivier/virtualenvs/toxtests/bin/tox', 'USER': 'olivier', 'HOME': '/home/olivier', 'OLDPWD': '/media/sf_CustomLibs/django-addanother/test_project', 'SSH_CONNECTION': '10.0.2.2 52110 10.0.2.15 22', 'LOGNAME': 'olivier', 'VIRTUAL_ENV': '/media/sf_CustomLibs/django-addanother/.tox/py34-django18', 'LS_COLORS': 'rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lz=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.axa=00;36:*.oga=00;36:*.spx=00;36:*.xspf=00;36:', 'LANG': 'en_US.UTF-8', 'LESSCLOSE': '/usr/bin/lesspipe %s %s', 'PYTHONHASHSEED': '1809464592', 'COMP_WORDBREAKS': ' \t\n"\'><;|&(:', 'LANGUAGE': 'en_US:en', 'PWD': '/media/sf_CustomLibs/django-addanother', 'PIP_ALLOW_EXTERNAL': 'true', 'LESSOPEN': '| /usr/bin/lesspipe %s', 'MAIL': '/var/mail/olivier', 'SSH_CLIENT': '10.0.2.2 52110 22', 'PATH': '/media/sf_CustomLibs/django-addanother/.tox/py34-django18/bin:/home/olivier/virtualenvs/toxtests/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games', 'SHLVL': '1', 'PS1': '(toxtests) \\[\\e]0;\\u@\\h: \\w\\a\\]${debian_chroot:+($debian_chroot)}\\u@\\h:\\w\\$ ', 'XDG_SESSION_ID': '6', 'TERM': 'xterm', 'XDG_RUNTIME_DIR': '/run/user/1000', 'DJANGO_SETTINGS_MODULE': 'test_project.settings'}

Using real prefix '/usr'
Traceback (most recent call last):
  File "/home/olivier/virtualenvs/toxtests/lib/python3.4/site-packages/virtualenv.py", line 349, in copyfile
    os.symlink(srcpath, dest)
OSError: [Errno 71] Protocol error: '/usr/lib/python3.4/config-3.4m-i386-linux-gnu' -> '/media/sf_CustomLibs/django-addanother/.tox/py34-django18/lib/python3.4/config-3.4m-i386-linux-gnu'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/olivier/virtualenvs/toxtests/lib/python3.4/site-packages/virtualenv.py", line 2332, in <module>
    main()
  File "/home/olivier/virtualenvs/toxtests/lib/python3.4/site-packages/virtualenv.py", line 711, in main
    symlink=options.symlink)
  File "/home/olivier/virtualenvs/toxtests/lib/python3.4/site-packages/virtualenv.py", line 924, in create_environment
    site_packages=site_packages, clear=clear, symlink=symlink))
  File "/home/olivier/virtualenvs/toxtests/lib/python3.4/site-packages/virtualenv.py", line 1129, in install_python
    copyfile(join(stdlib_dir, fn), join(lib_dir, fn), symlink)
  File "/home/olivier/virtualenvs/toxtests/lib/python3.4/site-packages/virtualenv.py", line 352, in copyfile
    copyfileordir(src, dest, symlink)
  File "/home/olivier/virtualenvs/toxtests/lib/python3.4/site-packages/virtualenv.py", line 327, in copyfileordir
    shutil.copytree(src, dest, symlink)
  File "/home/olivier/virtualenvs/toxtests/lib/python3.4/shutil.py", line 343, in copytree
    raise Error(errors)
shutil.Error: [('/home/olivier/virtualenvs/toxtests/lib/python3.4/config-3.4m-i386-linux-gnu/libpython3.4.so', '/media/sf_CustomLibs/django-addanother/.tox/py34-django18/lib/python3.4/config-3.4m-i386-linux-gnu/libpython3.4.so', "[Errno 71] Protocol error: '../../i386-linux-gnu/libpython3.4m.so.1' -> '/media/sf_CustomLibs/django-addanother/.tox/py34-django18/lib/python3.4/config-3.4m-i386-linux-gnu/libpython3.4.so'"), ('/home/olivier/virtualenvs/toxtests/lib/python3.4/config-3.4m-i386-linux-gnu/libpython3.4m.so', '/media/sf_CustomLibs/django-addanother/.tox/py34-django18/lib/python3.4/config-3.4m-i386-linux-gnu/libpython3.4m.so', "[Errno 71] Protocol error: '../../i386-linux-gnu/libpython3.4m.so.1' -> '/media/sf_CustomLibs/django-addanother/.tox/py34-django18/lib/python3.4/config-3.4m-i386-linux-gnu/libpython3.4m.so'")]
Running virtualenv with interpreter /home/olivier/virtualenvs/toxtests/bin/python3.4

```

**Comment by jonashaag** [comment]
WTF, no idea what's going on there. Never mind about tox, just run the test commands manually:

```
py.test -vv --cov django_addanother --create-db --strict -r fEsxXw test_project/tests.py
flake8 --show-source --exclude tests --max-complexity=5 --ignore=D203,E501,I101 django_addanother
```

**Comment by jonashaag** [comment]
I'll deal with the tox issue if it's not only a problem with your machine. Don't worry about that now. I don't want you to have to deal with such things while you're already working hard on contributing a feature (thanks a million btw!)

**Comment by olivierdalang** [comment]
I added jsi18n to the media and it works well now !

I'm very sorry about the tests though, I don't manage to run them here. I tried with the command directly, but then there seem to be a lot of missing dependencies (I stopped trying when I had to install firefox ;) )

I think it's normal they broke because I completely changed the test app, if you prefer I will remove everything regarding the test project from the PR, so that it should work. Still, I think we need a test project with foreign keys to showcase the edit link, and anyway as @jpic said, we'll need some tests for the new feature.
So either one of you takes care of this, or I'll have to dig a bit further in how to make those tests work :) BTW, I'm not familiar with testing at all, but wouldn't it be possible to have tests using `manage.py test` ? I had more success with this approach on other projects.

Thanks !

**Comment by jpic** [comment]
We use pytest-splinter, so you can switch browser with the `--browser` switch for pytest. Dependencies are listed in the tox.ini file.

We only have little functional tests so it should be pretty quick to change them to match the new demo app, since we **need** an FK to test the change feature because it doesn't work on M2M relations.

With manage.py test, you would not benefit from pytest, and neither would you benefit from pytest-splinter, which means you wouldn't be able to change the browser on the command line with `--browser`.

I really can't help right now, I've got too many friends waiting in line already ^^ Perhaps Jonas can, that said, again tests are small here, so it's IMHO the best project you could find to learn about pytest and splinter, you know, if you're into this kind of stuff xD

**Comment by jpic** [comment]
Could you perhaps open a separate issue and paste the output you get from the pytest command using the `--browser` of your choice so that we can try to help ? Thanks !!

**Comment by olivierdalang** [comment]
I could fix the tests in the meantime ;) I didn't manage to make them work locally but it's OK since travis builds are almost real-time.

I added a test for the new edit link.

It seems there's some change in the  `admin/popup_response.html` in django 1.10, leading to a failing test under 1.10. I'll try to fix that.

Besides, I had to remove reference to admin/jsi18n since it requires the user to be logged. I reimplemented the needed function (interpolate) in `static/django_addanother/django_jquery.js` instead. Would that be OK ? I'm know almost nothing about JS (especially namespacing), so it may be worth reviewing if the implementation is acceptable.

**Comment by jpic** [comment]
:guitar:

**Comment by olivierdalang** [comment]
Finally !

Thanks for the guidance and the support ;)

**Comment by jonashaag** [comment]
@olivierdalang Thanks so much for all the effort. I'll review this ASAP

File: django_addanother/templates/django_addanother/related_widget_wrapper.html (hunk 2 above)
Comment (jonashaag) [inline django_addanother/templates/django_addanother/related_widget_wrapper.html]: bump

File: django_addanother/templates/django_addanother/related_widget_wrapper.html [hunk 5]
```diff
@@ -2,9 +2,22 @@
...
+    <a class="related-widget-wrapper-link change-related" id="change_id_{{ name }}"
+        href=""
+        data-href-template="{{ edit_related_url }}?{{ url_params }}"
+        title="{% blocktrans %}Edit related{% endblocktrans %}">
+        <img src="{% static edit_icon %}" alt="{% trans 'Edit' %}"/>
+    </a>
+    {% endif %}
+
+    {% if add_related_url %}
+    &nbsp;
```
Comment (jonashaag) [inline django_addanother/templates/django_addanother/related_widget_wrapper.html]: same

File: django_addanother/views.py [hunk 6]
```diff
@@ -27,18 +31,23 @@ def form_valid(self, form):
...
+            return SimpleTemplateResponse('admin/popup_response.html', {
+                'action': self.POPUP_ACTION,
+                'value': six.text_type(self._get_created_obj_pk(created_obj)),
+                'obj': six.text_type(self.label_from_instance(created_obj)),
+                'new_value': six.text_type(self._get_created_obj_pk(created_obj)),
+            })
+        else:
+            return SimpleTemplateResponse('admin/popup_response.html', {
+                'popup_response_data':
+                    '{"action":"%s","value":"%s","obj":"%s","new_value":"%s"}' % (
```
Comment (jonashaag) [inline django_addanother/views.py]: Please use `json.dumps` here like in `django.contrib.admin`, which also saves you from using `escapejs`

File: django_addanother/views.py [hunk 7]
```diff
@@ -9,6 +10,9 @@ class PopupMixin(object):
     """Mixin for :class:`~django.views.generic.edit.CreateView` classes that
     handles the case of the view being opened in an *add another* popup window.
     """
+
+    POPUP_ACTION = 'add'
```
Comment (jonashaag) [inline django_addanother/views.py]: Factor this into three classes `BasePopupMixin`, `AddPopupMixin`, `ChangePopupMixin`

File: django_addanother/widgets.py [hunk 8]
```diff
@@ -1,15 +1,17 @@
...
 from django.utils.safestring import mark_safe
 
 
 if django.VERSION < (1, 9):
     DEFAULT_ADD_ICON = 'admin/img/icon_addlink.gif'
+    DEFAULT_EDIT_ICON = 'admin/img/icon_changelink.gif'
 else:
     DEFAULT_ADD_ICON = 'admin/img/icon-addlink.svg'
+    DEFAULT_EDIT_ICON = 'admin/img/icon-changelink.svg'
+DEFAULT_EDIT_ICON_DISABLED = 'django_addanother/img/icon-changelink-disabled.svg'
```
Comment (jonashaag) [inline django_addanother/widgets.py]: Why do we have an extra icon for disabled? Why not simply make it say 80% opacity?

File: test_project/testapp/models.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+from django.db import models
+
+class Team(models.Model):
+    name = models.CharField(max_length=20)
+    def __str__(self):
+        return self.name
+class Player(models.Model):
+    name = models.CharField(max_length=20)
+    current_team = models.ForeignKey("Team",related_name="current_players",help_text="This demonstrate the wrapper adding a create button only")
```
Comment (jonashaag) [inline test_project/testapp/models.py]: Your Python formatting is seriously broken in this file ;) Try `pylint` or `pep8`

File: test_project/testapp/templates/admin/popup_response_debug.html [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{% load i18n %}<!DOCTYPE html>
```
Comment (jonashaag) [inline test_project/testapp/templates/admin/popup_response_debug.html]: I think this file isn't used anywhere?

File: test_project/testapp/views.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    form = PlayerForm()
+
+    context = {
+        'form': form,
+        'jquery' : 'admin/js/jquery.js' if django.VERSION < (1, 9) else 'admin/js/vendor/jquery/jquery.js',
+    }    
+
+    return render(request, 'testapp/main.html', context)
+
+# class UpdatePlayer(generic.UpdateView):
```
Comment (jonashaag) [inline test_project/testapp/views.py]: Please remove if unused

File: test_project/testapp/views.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
+import django
+from django.views import generic
+from django_addanother.views import PopupMixin, ChangePopupMixin
+from .models import Team, Player
+from .forms import TeamForm, PlayerForm
+from django.core.urlresolvers import reverse_lazy
+from django.shortcuts import render
```
Comment (jonashaag) [inline test_project/testapp/views.py:3]: Please order these imports according to pep8:

> Imports should be grouped in the following order:
> 1. standard library imports
> 2. related third party imports
> 3. local application/library specific imports
> You should put a blank line between each group of imports.

File: test_project/testapp/views.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+class CreateTeam(PopupMixin, generic.CreateView):
+    model = Team
+    fields = ['name']
+
+class EditTeam(ChangePopupMixin, generic.UpdateView):
+    model = Team
+    form_class = TeamForm
+
+def main(request):
+    # Player = Player()
```
Comment (jonashaag) [inline test_project/testapp/views.py]: Please remove if unused

File: test_project/urls.py [hunk 14]
```diff
@@ -1,10 +1,12 @@
...
+from testapp.views import CreateTeam, EditTeam, main
 
-from views import CreateGroup, MainView
-
+from django import views as django_views
 urlpatterns = [
-    url(r'^admin/', admin.site.urls),
-    url(r'^groups/add/$', CreateGroup.as_view(), name='add_group'),
-    url(r'^$', MainView.as_view(), name='main'),
+	url(r'^admin/', admin.site.urls),
```
Comment (jonashaag) [inline test_project/urls.py]: Broken indentation here

File: django_addanother/views.py (hunk 7 above)
Comment (olivierdalang) [inline django_addanother/views.py]: This will break backwards compatibility, isn't this an issue ?

File: django_addanother/widgets.py (hunk 8 above)
Comment (olivierdalang) [inline django_addanother/widgets.py]: actually it's the case, but I forgot that line

**Comment by olivierdalang** [comment]
Ok I made the changes you asked for. Thanks for the detailed review.

As commented, factoring PopupMixin will break backwards compatibility (personally I don't mind).

**Comment by jonashaag** [comment]
Awesome! This looks real good now! Thanks for the backwards compatibility pointer, didn't think of that. I'd say never mind; @jpic what do you think? Should we provide a deprecation path?

File: django_addanother/templates/django_addanother/related_widget_wrapper.html (hunk 5 above)
Comment (jpic) [inline django_addanother/templates/django_addanother/related_widget_wrapper.html]: It probably makes in render better by default, I can live with it even though I'd prefer this in a separate commit.

File: django_addanother/views.py (hunk 7 above)
Comment (jpic) [inline django_addanother/views.py]: Backward compatibility with what ?

File: django_addanother/views.py (hunk 6 above)
Comment (jpic) [inline django_addanother/views.py]: Why not JsonResponse ?

File: django_addanother/widgets.py (hunk 8 above)
Comment (jpic) [inline django_addanother/widgets.py]: :guitar:

**Comment by jpic** [comment]
Perhaps it could just be documented and be breaked in a middle version increment ?

Is the history final or will it be squashed before merging ?

Also, <3 the patch and epic review by @jonashaag, can't wait to see that released, you :guitar: B)

File: django_addanother/views.py (hunk 7 above)
Comment (jpic) [inline django_addanother/views.py]: Not IMHO for a middle version increment if it's properly documented. We'll also need a paste of the traceback that happens when this package is upgraded without the renaming done in the project, for posterity in the documentation and google-fu treasure hunting action by other nerds like us LOL

@olivierdalang good thinking about the user community !

**Comment by jonashaag** [comment]
Merged! Thanks @olivierdalang for all the hard work!

I was just thinking, maybe we should rename the popups to `CreatePopupMixin` and `UpdatePopupMixin` to be in line with the Django generic class names?

**Comment by jonashaag** [comment]
@jpic I'm a bit lost with pytest, do you know what causes this:

```
RuntimeError: Model class test_project.testapp.models.Team doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.
```

I simply added a non-browser/unit test to the tests file but it won't import any models.

**Comment by jpic** [comment]
What's the complete traceback ?

What's the patch you simply added ?

**Comment by jonashaag** [comment]
TB: https://gist.github.com/jonashaag/5796423b720201fee74797c4b8e398d5
Patch:

```
+def test_smoke():
+    from testapp.models import Team, Player
+    print(Team.objects.all())
+    print(Player.objects.all())
```

**Comment by olivierdalang** [comment]
Awesome, thanks for the merge ! 

About the remaining issues ("rename the popups to CreatePopupMixin and UpdatePopupMixin"), do you want me to do it ?

**Comment by jonashaag** [comment]
> About the remaining issues ("rename the popups to CreatePopupMixin and UpdatePopupMixin"), do you want me to do it ?

I'll wait for feedback @jpic; also what's your opinion on that @olivierdalang?

**Comment by jpic** [comment]
Well perhaps we could do a major version bump and feel free to break BC, it's nice to have a consistent codebase.

We might as well rename the package, cause it's not only add-another anymore ... :dancer:

**Comment by jonashaag** [comment]
> We might as well rename the package, cause it's not only add-another anymore ... 💃

True, do you have a suggestions? django-editrelated? Oh we could add deletion as well and name it django-crudrelated :D

**Comment by jpic** [comment]
django-related-widget-wrapper perhaps ?​ For the module name, users prefer
to type acronyms even though I personnaly don't like them, related_ww or
just rww ?

More ideas ? :D
